batch-processing/pipeline.go

```diff
@@ -0,0 +1,354 @@
+package main
+
+import (
+	"crypto/md5"
+	"encoding/csv"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+const (
+	READ     = "read"
+	DOWNLOAD = "download"
+	CONVERT  = "convert"
+	UPLOAD   = "upload"
+	WRITE    = "write"
+)
+
+var TASKS = []string{DOWNLOAD, CONVERT, UPLOAD}
+
+var (
+	OutputHeader       = []string{"url", "input", "output", "s3url"}
+	FailedOutputHeader = []string{"url"}
+)
+
+var (
+	ErrDuplicateURL    = errors.New("duplicate URL, skipping")
+	ErrImageExistsInS3 = errors.New("image already exists in S3, skipping")
+)
+
+// Pipeline struct
+type Pipeline struct {
+	config        *Config
+	workers       map[string]int
+	maxRetries    uint64    // Max number of retries for downloading and uploading an image
+	urlMap        *sync.Map // this will record urls that are already processed so we don't download again.
+	tmpWorkingDir string
+
+	// channels
+	readOut     chan *ReadOut
+	downloadOut chan *DownloadOut
+	convertOut  chan *ConvertOut
+	uploadOut   chan *UploadOut
+	errOut      chan *ErrOut
+}
+
+func (p *Pipeline) Read(csvReader *csv.Reader, out chan *ReadOut) {
+	// Read the input CSV file
+	// For each line, send the URL to the download channel
+	defer close(out)
+
+	csvReader.FieldsPerRecord = 1
+
+	for {
+		row, err := csvReader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			url := ""
+			if len(row) > 0 {
+				url = row[0]
+			}
+			p.errOut <- &ErrOut{
+				Url: url,
+				Err: err,
+			}
+			continue
+		}
+
+		// check if URL was already read
+		if _, ok := p.urlMap.LoadOrStore(row[0], true); ok {
+			p.errOut <- &ErrOut{
+				Url: row[0],
+				Err: ErrDuplicateURL,
+			}
+			continue
+		}
+		out <- &ReadOut{Url: row[0]}
+	}
+	log.Println("Read done")
+}
+
+func (p *Pipeline) Download(wg *sync.WaitGroup, in <-chan *ReadOut, out chan *DownloadOut) {
+	// Download the image from the URL
+	// Send the image to the convert channel
+	defer wg.Done()
+	var (
+		hash string
+		ext  string
+	)
+	for row := range in {
+		md5 := md5.Sum([]byte(row.Url))
```

Comment:
So I decided to use URLs checksum as a temporary file name after I implemented the duplicate checking at read stage. 
When list of URLs has already been checked for duplicates and the user is aware that duplicates will be skipped - I think it's a good assumption to make. 

Otherwise, duplicate URLs could lead to files being replaced.